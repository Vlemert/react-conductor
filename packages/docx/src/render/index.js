import fs from 'fs';
import path from 'path';
import createElement from '../utils/createElement';
import WordRenderer from '../reconciler/';
import parse from '../parse/';

// Renders the input component
async function render(element, filePath) {
  // Create root container instance
  const container = createElement('ROOT');

  // Returns the current fiber (flushed fiber)
  const node = WordRenderer.createContainer(container);

  // Schedules a top level update with current fiber and a priority level (depending upon the context)
  WordRenderer.updateContainer(element, node, null);

  // Parse the input component and return the output
  const output = parse(container).toBuffer();

  // Officegen generates a output stream and not a file
  const stream = fs.createWriteStream(filePath);

  await new Promise((resolve, reject) => {
    // Generate a word document
    output.doc.generate(stream, Events(filePath, resolve, reject));
  });
}

// Handle docx events (optional)
function Events(filePath, resolve, reject) {
  return {
    finalize: () => {
      console.log(`âœ¨  Word document created at ${path.resolve(filePath)}.`);
      resolve();
    },
    error: () => {
      console.log('An error occurred while generating the document.');
      reject();
    },
  };
}

export default render;
