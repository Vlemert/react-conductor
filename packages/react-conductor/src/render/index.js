import { app } from 'electron';

import createElement from '../utils/create-element';
import ElectronRenderer from '../reconciler/';

// Renders the input component
async function render(element) {
  // Create root container instance
  const container = createElement('ROOT');

  // Returns the current fiber (flushed fiber)
  const node = ElectronRenderer.createContainer(container);

  // We wait for app to be ready before doing anything. This works for now, but
  // there might be a better way to do this
  app.once('ready', launchInfo => {
    // This is the only way to get `launchInfo` into the tree for now (we need
    // it for the `App.onReady` prop)
    container.launchInfo = launchInfo;

    // Schedules a top level update with current fiber and a priority level (depending upon the context)
    ElectronRenderer.updateContainer(element, node, null);
  });

  // TODO: figure out what to return here, ReactDOM does this:
  // return DOMRenderer.getPublicRootInstance(root);
}

export default render;
