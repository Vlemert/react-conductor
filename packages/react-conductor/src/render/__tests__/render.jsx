import React from 'react';
import { app, BrowserWindow } from 'electron';

import { Window } from '../../index';
import render from '../index';

jest.mock('electron');

/**
 * We should probably reorganize the tests. This no longer tests `Root`.
 */
describe('render', () => {
  beforeEach(() => {
    BrowserWindow.mockReset();
    app.once = jest.fn();
  });

  afterEach(() => {
    delete app.once;
  });

  async function testRender(element) {
    const renderPromise = render(element);
    app.once.mock.calls[0][1]();
    return renderPromise;
  }

  test('returns the component instance of the root element passed to it', async () => {
    class Application extends React.Component {
      render() {
        return <Window />;
      }
    }

    const renderResult = await testRender(<Application />);

    expect(renderResult).toBeInstanceOf(Application);
  });

  test('throws when trying to render a string', async () => {
    // Hide react error details from the console
    const originalConsoleError = console.error;
    console.error = jest.fn();

    const Application = () => 'hello';

    await expect(testRender(<Application />)).rejects.toMatchSnapshot();

    // And restore console.error behaviour
    console.error = originalConsoleError;
  });

  test('throws when trying to render an invalid component', async () => {
    // Hide react error details from the console
    const originalConsoleError = console.error;
    console.error = jest.fn();

    const Fake = 'fake';
    const Application = () => <Fake />;

    await expect(testRender(<Application />)).rejects.toMatchSnapshot();

    // And restore console.error behaviour
    console.error = originalConsoleError;
  });
});
