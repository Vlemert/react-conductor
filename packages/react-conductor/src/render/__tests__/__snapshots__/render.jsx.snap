// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render throws when trying to render a string 1`] = `[Error: TextElements are not supported! (do you have some text in your JSX?)]`;

exports[`render throws when trying to render an invalid component 1`] = `[Error: type 'fake' is not implemented (yet?)!]`;
